crates/store/re_datafusion/src/pushdown_expressions.rs

```diff
@@ -0,0 +1,412 @@
+        op,
+        right: Box::new(expr.left.as_ref().clone()),
+    }
+}
+
+pub(crate) fn filter_expr_is_supported(
+    filter_expr: &Expr,
+    query_dataset_request: &QueryDatasetRequest,
+    schema: &SchemaRef,
+) -> Result<TableProviderFilterPushDown, DataFusionError> {
+    let returned_queries =
+        apply_filter_expr_to_queries(vec![query_dataset_request.clone()], filter_expr, schema)?;
+
+    if returned_queries.is_some() {
+        Ok(TableProviderFilterPushDown::Inexact)
+    } else {
+        Ok(TableProviderFilterPushDown::Unsupported)
+    }
+}
+
+/// Apply a filter expression to a dataset query.
+/// This function will return Ok(None) if we cannot push down this filter into our request.
+/// It will return an error if the expression pushes down to return no results. This can
+/// occur if you have two mutually exclusive expressions that cannot overlap, such as
+/// `rerun_segment_id == "aaaa" AND rerun_segment_id == "BBBB"`.
+pub(crate) fn apply_filter_expr_to_queries(
+    queries: Vec<QueryDatasetRequest>,
+    expr: &Expr,
+    schema: &SchemaRef,
+) -> Result<Option<Vec<QueryDatasetRequest>>, DataFusionError> {
+    Ok(match expr {
+        Expr::Alias(alias_expr) => {
+            apply_filter_expr_to_queries(queries, alias_expr.expr.as_ref(), schema)?
+        }
+        Expr::BinaryExpr(expr) => {
+            let BinaryExpr { left, op, right } = arrange_binary_expr_as_col_on_left(expr);
+
+            match op {
+                Operator::And => {
+                    // When we have multiple queries they are effectively ORed together.
+                    // When we apply the expression to both the left and right we will
+                    // have (leftA OR leftB) AND (rightC OR rightD). We need to
+                    // consider the combinatorial for the final output.
+
+                    let Some(left_queries) =
+                        apply_filter_expr_to_queries(queries.clone(), &left, schema)?
+                    else {
+                        return Ok(None);
+                    };
+                    let Some(right_queries) =
+                        apply_filter_expr_to_queries(queries.clone(), &right, schema)?
+                    else {
+                        return Ok(None);
+                    };
+
+                    let final_exprs = left_queries
+                        .iter()
+                        .flat_map(|left| {
+                            right_queries
+                                .iter()
+                                .map(|right| merge_queries_and(left, right))
+                        })
+                        .collect::<Result<Vec<_>, _>>()?;
+
+                    Some(final_exprs)
+                }
+                Operator::Or => {
+                    let Some(mut left_queries) =
+                        apply_filter_expr_to_queries(queries.clone(), &left, schema)?
+                    else {
+                        return Ok(None);
+                    };
```

Comment:
I believe Jeremy's comment applies here too. In particular, the `InList` case yields a leading `lit(true)`, which I believe causes this to eroneously short-circuit.